### PR TITLE
Mss 783 extract notifications sent

### PR DESCRIPTION
**Pull request by alexduf**
This extracts report rows from our report dynamo DB table. It includes every field of the table, to be as thorough as possible.

This will allow us to have a record of each notification ever sent by this system in the datalake. I think the oldest record we have is July 2016, so I'll backdate from there.

### CODE REVIEW

File: reportextractor/src/main/scala/com/gu/notifications/extractor/Lambda.scala [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+import com.amazonaws.services.dynamodbv2.model._
+import com.amazonaws.services.dynamodbv2.{AmazonDynamoDB, AmazonDynamoDBClientBuilder}
+import com.amazonaws.services.lambda.runtime.{Context, RequestHandler}
+import com.amazonaws.services.s3.AmazonS3ClientBuilder
+import com.gu.{AppIdentity, AwsIdentity}
+import models.NotificationType
+import org.slf4j.{Logger, LoggerFactory}
+import play.api.libs.json.{JsValue, Json}
+import utils.MobileAwsCredentialsProvider
+
```
Comment (NathanielBennett) [inline reportextractor/src/main/scala/com/gu/notifications/extractor/Lambda.scala]: Think you've got a couple unused dependencies in here
AmazonDynamoDB, Duration, TemporalUnit

File: reportextractor/src/main/scala/com/gu/notifications/extractor/Lambda.scala [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+  }
+
+  val dynamoDB = AmazonDynamoDBClientBuilder.standard().withCredentials(credentials).withRegion(region).build()
+  val s3 = AmazonS3ClientBuilder.standard().withCredentials(credentials).withRegion(region).build()
+
+  val tableName: String = identity match {
+    case AwsIdentity(_, _, stage, _) => s"mobile-notifications-reports-$stage"
+    case _ => s"mobile-notifications-reports-CODE"
+  }
+
```
Comment (NathanielBennett) [inline reportextractor/src/main/scala/com/gu/notifications/extractor/Lambda.scala]: This seems much preferable to blowing cash on parameter store. :)

**Comment by NathanielBennett** [comment]
OK, sorry. Maybe two. Looks good though

File: reportextractor/cfn.yaml [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+                - !Sub arn:aws:dynamodb:${AWS::Region}:${AWS::AccountId}:table/mobile-notifications-reports-${Stage}/index/sentTime-index
+        - PolicyName: s3
+          PolicyDocument:
+            Statement:
+              Effect: Allow
+              Action:
+                - "s3:PutObject"
+              Resource:
+                - !Ref OutputBucket
+                - !Sub ${OutputBucket}/*
```
[the remark this answers is not in the document]
Comment (alexduf) [reply reportextractor/cfn.yaml]: very much so!

File: reportextractor/cfn.yaml [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+        - PolicyName: logs
+          PolicyDocument:
+            Statement:
+              Effect: Allow
+              Action:
+                - logs:CreateLogGroup
+                - logs:CreateLogStream
+                - logs:PutLogEvents
+                - cloudwatch:putMetricData
+              Resource: "*"
```
[the remark this answers is not in the document]
Comment (alexduf) [reply reportextractor/cfn.yaml]: Good shout

File: reportextractor/src/main/scala/com/gu/notifications/extractor/Lambda.scala (hunk 1 above)
Comment (alexduf) [reply reportextractor/src/main/scala/com/gu/notifications/extractor/Lambda.scala]: Yep, good eye here!

**Pull request merged**